li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -208,7 +230,85 @@ private LiKafkaFederatedProducerImpl(LiKafkaProducerConfig configs,
 
   @Override
   public Future<RecordMetadata> send(ProducerRecord<K, V> producerRecord, Callback callback) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    String topic = producerRecord.topic();
+    PerClusterProducerContext context = getOrCreateProducerContextForTopic(topic);
+    Producer<byte[], byte[]> producer = context.producer();
+    Auditor<K, V> auditor = context.auditor();
+    LongAdder numThreadsInSend = context.numThreadsInSend();
+
+    numThreadsInSend.increment();
+    boolean failed = true;
+    try {
+      if (context.closed()) {
+        throw new IllegalStateException("LiKafkaProducer has been closed.");
+      }
+
+      K key = producerRecord.key();
+      V value = producerRecord.value();
+      Object auditToken = auditor.auditToken(key, value);
+      Long timestamp = producerRecord.timestamp() == null ? System.currentTimeMillis() : producerRecord.timestamp();
+      Integer partition = producerRecord.partition();
+      Future<RecordMetadata> future = null;
+      UUID messageId = _uuidFactory.getUuid(producerRecord);
+      if (LOG.isTraceEnabled()) {
+        LOG.trace("Sending event: [{}, {}] with key {} to kafka topic {} to cluster {}",
+            messageId.toString().replaceAll("-", ""),
+            value.toString(),
+            (key != null) ? key.toString() : "[none]",
+            topic,
+            context.cluster().name());
+      }
+      byte[] serializedValue;
+      byte[] serializedKey;
+      try {
+        serializedValue = _valueSerializer.serialize(topic, value);
+        serializedKey = _keySerializer.serialize(topic, key);
+      } catch (Throwable t) {
+        // Audit the attempt and the failure.
+        auditor.record(auditToken, topic, timestamp, 1L, 0L, AuditType.ATTEMPT);
+        throw t;
+      }
+      int sizeInBytes = (serializedKey == null ? 0 : serializedKey.length)
```

Comment:
do we want to throw in headers into the size calculation ?

Reply:
I'll leave it as it is because I think other code assumes that header size is not included in the audited size.